kernel/objects/signal-listener/objects/SignalListener.cc

```diff
@@ -0,0 +1,250 @@
+    else if (data.keventSink() != null_cap) {
+      auto err = setSink(msg->lookupEntry(data.keventSink()));
+      if (!err) return err.state();
+
+    }
+    _mutex << [this, &data] {
+      _signal &= ~data.resetMask;
+      // Comment on races after bind of kevent sink:
+      // read cap fresh from _sink entry because of race with unbind after bind
+      // if it was called, the cap is killed now (-> not usable)
+      // we can not race with another binding call,
+      // because this is protected by the monitor
+      TypedCap<IKEventSink> sink(_sink);
+      if (!sink) return;
+      if (_signal && !_eventAttached) {
+        sink->attachKEvent(&_eventHandle);
+        _eventAttached = true;
+      } else if (!_signal && _eventAttached) {
+        sink->detachKEvent(&_eventHandle);
+        _eventAttached = false;
+      }
+    };
+
+    return Error::SUCCESS;
+  }
+
+  Error SignalListener::invokeReset(Tasklet*, Cap, IInvocation* msg)
+  {
+    auto data = msg->getMessage()->read<protocol::SignalListener::Reset>();
+    _mutex << [this, &data] {
+      _signal &= ~data.resetMask;
+      TypedCap<IKEventSink> sink(_sink);
+      if (!sink) return;
+      if (_signal && !_eventAttached) {
+        sink->attachKEvent(&_eventHandle);
+        _eventAttached = true;
+      } else if (!_signal && _eventAttached) {
+        sink->detachKEvent(&_eventHandle);
+        _eventAttached = false;
+      }
+    };
+    return Error::SUCCESS;
+  }
+
+
+  optional<void> SignalListener::setSource(optional<CapEntry*> entry)
+  {
+    TypedCap<ISignalSource> src(entry);
+    if (!src) RETHROW(src);
+    RETURN(_source.set(this, *entry, src.cap()));
+  }
+
+  void SignalListener::bind(optional<ISignalSource*> source)
+  {
+    ASSERT(source);
+    source->attachSignalSink(&_listenerHandle);
+  }
+
+  void SignalListener::unbind(optional<ISignalSource*> source)
+  {
+    ASSERT(source);
+    source->detachSignalSink(&_listenerHandle);
+  }
+
+  optional<void> SignalListener::setSink(optional<CapEntry*> entry)
+  {
+    TypedCap<IKEventSink> sink(entry);
+    if (!sink) RETHROW(sink);
+    RETURN(_sink.set(this, *entry, sink.cap()));
+  }
+
+  void SignalListener::bind(optional<IKEventSink*> sink)
+  {
+    ASSERT(sink);
+    // this function is called BEFORE the cap is commited into the CapRef in CapEntry::insertAfter
+    // we can not attach to _sink here because of races with signal changed
+  }
+
+  void SignalListener::unbind(optional<IKEventSink*> sink)
+  {
+    // this function is called AFTER the cap is killed in resetReference()
+    // and before it is reset
+    ASSERT(sink);
+    _mutex << [this, sink] {
+      sink->detachKEvent(&_eventHandle);
+      _eventAttached = false;
+    };
+  }
+
+  KEvent SignalListener::deliverKEvent()
+  {
+    KEvent result;
+    _mutex << [this, &result] {
+      result.user = _context;
+      result.state = _signal;
+
+      _eventAttached = false;
+      _signal &= ~_autoResetMask;
+
+      if (_signal) {
```

Comment:
we should not reemit the event until something changes in the signal state. for example store the last delivered signal, only redeliver if new signal has more bits on than previously delivered signal. otherwise the receiver gets bombarded with repeated.

but actually we don't need this filter functionality here because the signal source will only reemit the same event after the original cause was resolved. for example acknowledgement of an interrupt to unmask it, and also trap handling in an execution context. Therefore it is sufficient to reset the signal to 0 after delivery.

therefore this block and the reset call are not needed!?